pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,148 @@
+// All the Reserved IP ranges not present in the cloud provider are extracted from the reserved list in the IPAllocator
+// Finally a final reserved list is created by merging these two lists
+// Potential error cases: 1) No /29 Block in the CIDR is unreserved
+// 2) Parsing the CIDR resulted in an error
+func (ipAllocator *IPAllocator) GetUnreservedIPBlock(cidr string, reservedIPRanges map[string]bool) (string, error) {
+	// Merging the reserved lists at the cloud provider end and the service end to form a final reserved list
+	for reservedIPRange := range ipAllocator.reserved {
+		reservedIPRanges[reservedIPRange] = true
+	}
+
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("Unable to parse CIDR %s", cidr)
+	}
+	for cidrIPBlock := cloneIP(ip.Mask(ipnet.Mask)); ipnet.Contains(cidrIPBlock) && err == nil; err = incrementIP(cidrIPBlock, incrementStep29IPBlock) {
+		overLap := false
+		ipRange := fmt.Sprint(cidrIPBlock.String(), reservationBlockSize)
+		for reservedIPRange := range reservedIPRanges {
+			// Find if the current IP range in the CIDR overlaps with any of the reserved IP ranges. If not, this can be returned
+			overLap, err = isOverlap(ipRange, reservedIPRange)
+			// Error while processing cidr is also considered as an overlap to ignore the specific IP block
+			if overLap || err != nil {
+				break
+			}
+		}
+		if !overLap {
+			return ipRange, nil
+		}
+	}
+
+	// No unreserved IP range available in the entire CIDR range. Returning error
+	return "", fmt.Errorf("All of the /29 IP ranges in the cidr %s are reserved", cidr)
+}
+
+// isOverlap checks if two cidrs have any overlapping IPs
+func isOverlap(cidr1 string, cidr2 string) (bool, error) {
+	_, ipnet1, err := net.ParseCIDR(cidr1)
+	// Invalid CIDRs are considered as overlapping
+	if err != nil {
+		return true, fmt.Errorf("Invalid cidr %s provided", cidr1)
+	}
+
+	_, ipnet2, err := net.ParseCIDR(cidr2)
+	if err != nil {
+		return true, fmt.Errorf("Invalid cidr %s provided", cidr2)
+	}
+
+	return ipnet1.Contains(ipnet2.IP) || ipnet2.Contains(ipnet1.IP), nil
+}
+
+// Increment the given IP value by the provided step
+func incrementIP(ip net.IP, step byte) error {
+	ip = ip.To4()
+	for ipByte := 3; ipByte >= 0; ipByte-- {
+		//	If it the least significant byte, add step to it
+		if ipByte == 3 {
+			ip[ipByte] += step
+			// Addition did not result in a carry. We have the incremented value
+			if ip[ipByte] >= step {
+				break
+			}
+		} else {
+			// Add the carry bit from the previous addition. Value of carry can only be 1 as we are performing byte addition
+			ip[ipByte]++
+
+			// As the rollover case is 255->0, we conclude that addition did not result in a carry. We have the incremented value
+			if ip[ipByte] > 0 {
+				break
+			}
+			//We are at the most significant byte and adding 1 to it resulted in overflowing its value. IP Address rolled over. Returning error
```

Comment:
why is this an error? can't we then increment the next value by 1? 

`0.255.255.255` step by `1` should be `1.0.0.0` not an error

Reply:
this check is after incrementing ip[ipByte]. So, after increment, if this value is still 0, that implies that it transitioned from 255 to 0 and there is an overflow. If you are already at the Most Significant byte(ipByte== 0) and the value is 0 after increment, it means that there is an overflow in the IP range.